### PR TITLE
Misc visual studio cl support

### PR DESCRIPTION
**Pull request by krocard**
Various fixes for windows.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/01org/parameter-framework/pull/175%23issuecomment-136433181%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335160%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335464%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335687%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38336045%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38336299%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38421214%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38421306%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38421308%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38421338%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/175%23issuecomment-136433181%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3A%2B1%3A%20%40dawagner%20%40OznOg%20Please%20review%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A17%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20a253b9ac838b63b50eff0208deb879db6585fe8b%20bindings/c/ParameterFramework.h%2018%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335687%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22fallback%20to%20empty%20macros%20%3F%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A28%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%2C%20I%20want%20a%20new%20complicator%20to%20add%20the%20relevant%20macro.%20Not%20to%20fail%20silently.%20Anyway%20their%20is%20no%20req%20to%20support%20other%20compilator.%22%2C%20%22created_at%22%3A%20%222015-09-01T14%3A00%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-09-01T14%3A01%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20bindings/c/ParameterFramework.h%3AL49-68%22%7D%2C%20%22Pull%20a253b9ac838b63b50eff0208deb879db6585fe8b%20CMakeLists.txt%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335160%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Later%20options%20don%27t%20override%20former%20ones%20%3F%3F%3F%5Cr%5Cn%5Cr%5Cn%2A%2Aedit%2A%2A%3A%20I%20just%20read%20you%20commit%20message...%20that%27s%20too%20bad.%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A23%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A34%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CMakeLists.txt%3AL42-58%22%7D%2C%20%22Pull%20a253b9ac838b63b50eff0208deb879db6585fe8b%20utility/NonCopiable.hpp%2045%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38335464%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22although%20%5C%22copiable%5C%22%20seems%20to%20exist%2C%20%5C%22copyable%5C%22%20is%20more%20common.%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A26%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-09-01T14%3A01%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20utility/NonCopiable.hpp%3AL1-56%22%7D%2C%20%22Pull%20a253b9ac838b63b50eff0208deb879db6585fe8b%20xmlserializer/CMakeLists.txt%2011%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/175%23discussion_r38336045%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22fix%20the%20check%20instead.%22%2C%20%22created_at%22%3A%20%222015-08-31T17%3A32%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6430928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dawagner%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-09-01T14%3A01%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20xmlserializer/CMakeLists.txt%3AL44-57%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/175#issuecomment-136433181'>General Comment</a></b>
- <a href='https://github.com/krocard'></a> :+1: @dawagner @OznOg Please review
- [x] <a href='#crh-comment-Pull a253b9ac838b63b50eff0208deb879db6585fe8b CMakeLists.txt 7'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/175#discussion_r38335160'>File: CMakeLists.txt:L42-58</a></b>
- <a href='https://github.com/dawagner'></a> Later options don't override former ones ???
**edit**: I just read you commit message... that's too bad. :+1:
- [x] <a href='#crh-comment-Pull a253b9ac838b63b50eff0208deb879db6585fe8b utility/NonCopiable.hpp 45'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/175#discussion_r38335464'>File: utility/NonCopiable.hpp:L1-56</a></b>
- <a href='https://github.com/dawagner'></a> although "copiable" seems to exist, "copyable" is more common.
- [x] <a href='#crh-comment-Pull a253b9ac838b63b50eff0208deb879db6585fe8b bindings/c/ParameterFramework.h 18'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/175#discussion_r38335687'>File: bindings/c/ParameterFramework.h:L49-68</a></b>
- <a href='https://github.com/dawagner'></a> fallback to empty macros ?
- <a href='https://github.com/krocard'></a> No, I want a new complicator to add the relevant macro. Not to fail silently. Anyway their is no req to support other compilator.
- [x] <a href='#crh-comment-Pull a253b9ac838b63b50eff0208deb879db6585fe8b xmlserializer/CMakeLists.txt 11'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/175#discussion_r38336045'>File: xmlserializer/CMakeLists.txt:L44-57</a></b>
- <a href='https://github.com/dawagner'></a> fix the check instead.

<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/175?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/175?approve=1'></a>&nbsp;<a href='https://github.com/01org/parameter-framework/pull/175'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dawagner** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/01org/parameter-framework/pull/175'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by krocard** [comment]
:+1: @dawagner @OznOg Please review

File: CMakeLists.txt [hunk 1]
```diff
@@ -42,7 +42,16 @@ execute_process(COMMAND git describe --tags --dirty
     OUTPUT_VARIABLE PARAMETER_FRAMEWORK_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE)
 
 if (WIN32)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4 /WX")
+    # By default cmake adds a warning level.
+    # Delete it to force an other level.
+    STRING(REGEX REPLACE "/W[0-4]" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
```
Comment (dawagner) [inline CMakeLists.txt]: Later options don't override former ones ???

File: utility/NonCopiable.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * Sometime implementing copy/move semantic is complicated or not useful.
+ * The traditional way is to declare the relevant methods private in c++ < 11
+ * or deleted in c++ > 11.
+ *
+ * It is easier and self documented to derive from this class.
+ *
+ * For more documentation see:
+ * http://www.boost.org/doc/libs/release/libs/utility/utility.htm#Class_noncopyable
+ * */
+class NonCopiable
```
Comment (dawagner) [inline utility/NonCopiable.hpp]: although "copiable" seems to exist, "copyable" is more common.

File: bindings/c/ParameterFramework.h [hunk 3]
```diff
@@ -49,9 +49,19 @@ extern "C" {
...
+#    define NONNULL_(...)  __attribute__((nonnull (__VA_ARGS__)))
+#    define USERESULT __attribute__((warn_unused_result))
+#elif defined(_MSC_VER)
+    // In visual studio's cl there is no
+    // equivalent of nonnull
+#    define NONNULL
+#    define NONNULL_(...)
+#    define USERESULT _Check_return_
+#else
+#   error "Unknown compilator"
```
Comment (dawagner) [inline bindings/c/ParameterFramework.h]: fallback to empty macros ?

File: xmlserializer/CMakeLists.txt [hunk 4]
```diff
@@ -44,11 +44,13 @@ else(NOT LIBXML2_FOUND)
...
     include(CheckLibraryExists)
     check_library_exists(xml2 xmlXIncludeProcess "" XML2_XINCLUDE_SUPPORT)
-    if(NOT XML2_XINCLUDE_SUPPORT)
-        message(SEND_ERROR "
-    libxml2 has not been built with support for XInclude.  xmlserializer needs
-    that feature; please install a version of libxml2 supporting it.")
-    endif(NOT XML2_XINCLUDE_SUPPORT)
+
+    # Fixme: This check does not work on windows
+    if(NOT XML2_XINCLUDE_SUPPORT AND NOT WIN32)
```
Comment (dawagner) [inline xmlserializer/CMakeLists.txt]: fix the check instead.

File: CMakeLists.txt (hunk 1 above)
Comment (dawagner) [inline CMakeLists.txt]: :+1:

File: bindings/c/ParameterFramework.h (hunk 3 above)
Comment (krocard) [inline bindings/c/ParameterFramework.h]: No, I want a new complicator to add the relevant macro. Not to fail silently.

File: xmlserializer/CMakeLists.txt (hunk 4 above)
Comment (krocard) [inline xmlserializer/CMakeLists.txt]: :+1:<a href='#crh-update-none'></a>

File: utility/NonCopiable.hpp (hunk 2 above)
Comment (krocard) [inline utility/NonCopiable.hpp]: :+1:<a href='#crh-update-complete'></a>

File: bindings/c/ParameterFramework.h (hunk 3 above)
Comment (krocard) [inline bindings/c/ParameterFramework.h]: :+1:

**Pull request merged**